exchange/utils_test.go

```diff
@@ -1,11 +1,123 @@
+//
+// It always disallows BrightRoll for GDPR consent
+type permissionsMock struct{}
+
+func (p *permissionsMock) HostCookiesAllowed(ctx context.Context, consent string) (bool, error) {
+	return true, nil
+}
+
+func (p *permissionsMock) BidderSyncAllowed(ctx context.Context, bidder openrtb_ext.BidderName, consent string) (bool, error) {
+	return true, nil
+}
+
+func (p *permissionsMock) PersonalInfoAllowed(ctx context.Context, bidder openrtb_ext.BidderName, consent string) (bool, error) {
+	if bidder == "brightroll" {
+		return false, nil
+	}
+	return true, nil
+}
+
+func assertReqWithoutAliases(t *testing.T, reqByBidders map[openrtb_ext.BidderName]*openrtb.BidRequest) {
+	// assert individual bidder requests
+	assert.NotEqual(t, len(reqByBidders), 0, "cleanOpenRTBRequest should split request into individual bidder requests")
+
+	// assert for PI data
+	assert.Equal(t, reqByBidders["brightroll"].User.BuyerUID, "", "cleanOpenRTBRequest should clean PI data for a non-consented vendor")
+	assert.Equal(t, reqByBidders["brightroll"].Device.DIDMD5, "", "cleanOpenRTBRequest should clean PI data for a non-consented vendor")
+	assert.NotEqual(t, reqByBidders["appnexus"].User.BuyerUID, "", "cleanOpenRTBRequest shouldn't clean PI data for a consented vendor")
+}
+
+func assertReqWithAliases(t *testing.T, reqByBidders map[openrtb_ext.BidderName]*openrtb.BidRequest) {
+	// assert individual bidder requests
+	assert.NotEqual(t, reqByBidders, 0, "cleanOpenRTBRequest should split request into individual bidder requests")
+
+	// assert for PI data
+	assert.NotEqual(t, reqByBidders["brightroll"].User.BuyerUID, "", "cleanOpenRTBRequest shouldn't clean PI data for a consented vendor")
+	assert.NotEqual(t, reqByBidders["brightroll"].Device.DIDMD5, "", "cleanOpenRTBRequest shouldn't clean PI data for a consented vendor")
+	assert.NotEqual(t, reqByBidders["appnexus"].User.BuyerUID, "", "cleanOpenRTBRequest shouldn't clean PI data for a consented vendor")
+}
+
+func TestCleanOpenRTBRequests(t *testing.T) {
+	testCases := []struct {
+		req              *openrtb.BidRequest
+		bidReqAssertions func(t *testing.T, reqByBidders map[openrtb_ext.BidderName]*openrtb.BidRequest)
+		hasError         bool
+	}{
+		{req: newRaceCheckingRequest(t), bidReqAssertions: assertReqWithoutAliases, hasError: false},
+		{req: newAdapterAliasBidRequest(t), bidReqAssertions: assertReqWithAliases, hasError: false},
+	}
+
+	for _, test := range testCases {
+		reqByBidders, _, err := cleanOpenRTBRequests(context.Background(), test.req, &emptyUsersync{}, map[openrtb_ext.BidderName]*pbsmetrics.AdapterLabels{}, pbsmetrics.Labels{}, &permissionsMock{}, true)
+		if test.hasError {
+			assert.NotNil(t, err, "Error shouldn't be nil")
+		} else {
+			assert.Nil(t, err, "Err should be nil")
+			test.bidReqAssertions(t, reqByBidders)
+		}
+	}
+}
+
+// newAdapterAliasBidRequest builds a BidRequest with aliases
+func newAdapterAliasBidRequest(t *testing.T) *openrtb.BidRequest {
```

Comment:
We've a couple such requests and responses - some in the test functions and some as part of helper functions like this one - that we build and use for test cases in this package and I was wondering if we should put them all together in separate files like `sample_requests_test.go` and `sample_responses_test.go` (and maybe also have a dynamic req/res builder type function) for better organization and readability. But, this might require quite a bit of refactoring and therefore would like to know your thoughts @hhhjort @dbemiller @josephveach @mkendall07 @guscarreon before diving into it.

Reply:
The adapter test suite is probably the most feature rich and organized suite. It also benefits from all adapters being called through the same API, so you only need to run the test suite through the API.

On the exchange side it would be nice to have something similar. Unfortunately in the exchange loop, the adapter handlers will make calls out to the bidders to get bids back. To have proper exchange tests, we would need to mock out all these bidder responses, and ofcou

Reply:
There are a bunch of valid input requests here: https://github.com/prebid/prebid-server/tree/master/endpoints/openrtb2/sample-requests/valid-whole

They currently serve two purposes:

1. There's a test in `endpoints`  which makes sure that they all pass input validation.
2. I had hoped someday to generate user-facing docs from the `exemplary` ones. Like... "here's the  prototypical  way to build a  request using feature X." The `supplementary` ones wouldn't make good user-facing docs, but they're valid nonetheless, and exist to help fill out test coverage. Never got around to making that happen, thoughh :(.

Nothing preventing you from using them in other tests too, though!

Reply:
@dbemiller Yes, we can definitely use those json requests in this or other tests. I didn't really want to use that here because a bid request object made much more sense to me for this particular test case. 

Also, I was actually talking about the request/response objects that we define directly in our tests for some test cases. Like I am doing here with the `newAdapterAliasBidRequest` function. This test case is also leveraging another sample request defined by the `newRaceCheckingRequest` function. There are a couple other request/response objects in this package which we can potentially pull out and put together in corresponding `sample_requests_test.go` and `sample_responses_test.go` files, for example. Quite similar to how we have bundled the json request and responses together. Although, I think that will require some amount of refactoring and we might not want to do that as part of this PR.

Reply:
yeah, for sure. I wouldn't refactor all that in this PR either.

Just for some context... those `sample-requests` got added to the project fairly late, relative to all the tests with hardcoded objects. If I were starting from scratch I think I would have started with JSON and stuck with it.

The fact is that hardly any tests actually care about all those properties. They generally just want "any valid request... but let me set these 3-4 properties so I know what I'm working with." Since the `sample-requests` payloads run through other input validation tests, they're guaranteed to be valid. If you just define an object in a test file somewhere, the tests which use it are less focused. They might fail because there was a breaking API change and the object in the test wasn't valid anymore.... or because the object always had an invalid property, but none of the other tests used it in such a way that it mattered to them yet.

Just my 2 cents... not a strong opinion, and definitely not something I'd suggest blocking this PR to do.